File: app/csharp/AccountingSystem.Tests/Repositories/AuditLogRepositoryTest.cs
using AccountingSystem.Domain.Audit;
using AccountingSystem.Infrastructure.Persistence.Dapper.Entities;
using AccountingSystem.Infrastructure.Persistence.Repositories;
using FluentAssertions;
using Xunit;

namespace AccountingSystem.Tests.Repositories;

/// <summary>
/// AuditLogRepository の統合テスト
/// </summary>
public class AuditLogRepositoryTest : DatabaseTestBase
{
    [Fact]
    public async Task InsertAsync_ShouldInsertAuditLogAndReturnWithId()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);
        var auditLog = AuditLog.Create(
            "Account",
            "1100",
            AuditAction.CREATE,
            "user123",
            "山田太郎",
            new Dictionary<string, object>
            {
                ["code"] = "1100",
                ["name"] = "現金"
            },
            "192.168.1.1"
        );

        // Act
        var result = await repository.InsertAsync(auditLog);

        // Assert
        result.Should().NotBeNull();
        result.Id.Should().BeGreaterThan(0);
        result.EntityType.Should().Be("Account");
        result.EntityId.Should().Be("1100");
        result.Action.Should().Be(AuditAction.CREATE);
        result.UserId.Should().Be("user123");
        result.UserName.Should().Be("山田太郎");
    }

    [Fact]
    public async Task FindByEntityAsync_ShouldReturnAuditLogsForEntity()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);

        // 複数の監査ログを登録
        var log1 = AuditLog.Create("Account", "1100", AuditAction.CREATE, "user1", "ユーザー1",
            new Dictionary<string, object> { ["name"] = "現金" }, null);
        var log2 = AuditLog.CreateForUpdate("Account", "1100", "user2", "ユーザー2",
            new Dictionary<string, object> { ["name"] = "現金" },
            new Dictionary<string, object> { ["name"] = "小口現金" }, null);
        var log3 = AuditLog.Create("Account", "1200", AuditAction.CREATE, "user1", "ユーザー1",
            new Dictionary<string, object> { ["name"] = "普通預金" }, null);

        await repository.InsertAsync(log1);
        await repository.InsertAsync(log2);
        await repository.InsertAsync(log3);

        // Act
        var results = await repository.FindByEntityAsync("Account", "1100");

        // Assert
        results.Should().HaveCount(2);
        results.Should().AllSatisfy(log => log.EntityId.Should().Be("1100"));
        results.First().Action.Should().Be(AuditAction.UPDATE); // 最新順
        results.Last().Action.Should().Be(AuditAction.CREATE);
    }

    [Fact]
    public async Task FindByUserAsync_ShouldReturnAuditLogsForUser()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);
        var now = DateTime.UtcNow;

        var log1 = AuditLog.Create("Account", "1100", AuditAction.CREATE, "user123", "山田太郎",
            new Dictionary<string, object>(), null);
        var log2 = AuditLog.Create("Journal", "J2024-0001", AuditAction.CREATE, "user123", "山田太郎",
            new Dictionary<string, object>(), null);
        var log3 = AuditLog.Create("Account", "1200", AuditAction.CREATE, "user456", "佐藤次郎",
            new Dictionary<string, object>(), null);

        await repository.InsertAsync(log1);
        await repository.InsertAsync(log2);
        await repository.InsertAsync(log3);

        // Act
        var results = await repository.FindByUserAsync("user123", now.AddMinutes(-1), now.AddMinutes(1));

        // Assert
        results.Should().HaveCount(2);
        results.Should().AllSatisfy(log => log.UserId.Should().Be("user123"));
    }

    [Fact]
    public async Task FindByPeriodAsync_ShouldReturnAuditLogsWithinPeriod()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);
        var now = DateTime.UtcNow;

        // 複数ログを登録
        for (int i = 0; i < 5; i++)
        {
            var log = AuditLog.Create("Account", $"110{i}", AuditAction.CREATE, "user1", "ユーザー1",
                new Dictionary<string, object>(), null);
            await repository.InsertAsync(log);
        }

        // Act
        var results = await repository.FindByPeriodAsync(now.AddMinutes(-1), now.AddMinutes(1), 3);

        // Assert
        results.Should().HaveCount(3); // limit=3
    }

    [Fact]
    public async Task FindByActionAsync_ShouldReturnAuditLogsByAction()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);
        var now = DateTime.UtcNow;

        var createLog = AuditLog.Create("Account", "1100", AuditAction.CREATE, "user1", "ユーザー1",
            new Dictionary<string, object>(), null);
        var updateLog = AuditLog.CreateForUpdate("Account", "1100", "user2", "ユーザー2",
            new Dictionary<string, object>(), new Dictionary<string, object>(), null);
        var deleteLog = AuditLog.CreateForDelete("Account", "1200", "user1", "ユーザー1",
            new Dictionary<string, object>(), "テスト削除", null);

        await repository.InsertAsync(createLog);
        await repository.InsertAsync(updateLog);
        await repository.InsertAsync(deleteLog);

        // Act
        var createResults = await repository.FindByActionAsync(AuditAction.CREATE, now.AddMinutes(-1), now.AddMinutes(1), 100);
        var updateResults = await repository.FindByActionAsync(AuditAction.UPDATE, now.AddMinutes(-1), now.AddMinutes(1), 100);
        var deleteResults = await repository.FindByActionAsync(AuditAction.DELETE, now.AddMinutes(-1), now.AddMinutes(1), 100);

        // Assert
        createResults.Should().HaveCount(1);
        updateResults.Should().HaveCount(1);
        deleteResults.Should().HaveCount(1);
    }

    [Fact]
    public async Task InsertAsync_ShouldPersistJsonData()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);
        var oldValues = new Dictionary<string, object>
        {
            ["name"] = "現金",
            ["balance"] = 100000
        };
        var newValues = new Dictionary<string, object>
        {
            ["name"] = "小口現金",
            ["balance"] = 50000
        };

        var auditLog = AuditLog.CreateForUpdate(
            "Account", "1100", "user1", "ユーザー1",
            oldValues, newValues, "192.168.1.1"
        );

        // Act
        await repository.InsertAsync(auditLog);
        var results = await repository.FindByEntityAsync("Account", "1100");

        // Assert
        results.Should().HaveCount(1);
        var result = results.First();
        result.OldValues.Should().NotBeNull();
        result.NewValues.Should().NotBeNull();
    }

    [Fact]
    public async Task InsertAsync_ShouldPersistReason()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);
        var auditLog = AuditLog.CreateForDelete(
            "Account", "1100", "user1", "ユーザー1",
            new Dictionary<string, object> { ["name"] = "現金" },
            "重複登録のため削除",
            "192.168.1.1"
        );

        // Act
        await repository.InsertAsync(auditLog);
        var results = await repository.FindByEntityAsync("Account", "1100");

        // Assert
        results.Should().HaveCount(1);
        results.First().Reason.Should().Be("重複登録のため削除");
    }

    [Fact]
    public async Task FindByEntityAsync_ShouldReturnEmptyListWhenNoLogs()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);

        // Act
        var results = await repository.FindByEntityAsync("NonExistent", "000");

        // Assert
        results.Should().BeEmpty();
    }

    [Fact]
    public async Task FindByUserAsync_ShouldReturnEmptyListWhenNoLogs()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);
        var now = DateTime.UtcNow;

        // Act
        var results = await repository.FindByUserAsync("nonexistent", now.AddDays(-1), now);

        // Assert
        results.Should().BeEmpty();
    }

    [Fact]
    public async Task FindByEntityAsync_ShouldReturnLogsInDescendingOrder()
    {
        // Arrange
        var repository = new AuditLogRepository(ConnectionString);

        var log1 = AuditLog.Create("Account", "1100", AuditAction.CREATE, "user1", "ユーザー1",
            new Dictionary<string, object>(), null);
        await repository.InsertAsync(log1);
        await Task.Delay(10); // 時間差を確保

        var log2 = AuditLog.CreateForUpdate("Account", "1100", "user1", "ユーザー1",
            new Dictionary<string, object>(), new Dictionary<string, object>(), null);
        await repository.InsertAsync(log2);
        await Task.Delay(10);

        var log3 = AuditLog.CreateForDelete("Account", "1100", "user1", "ユーザー1",
            new Dictionary<string, object>(), "削除", null);
        await repository.InsertAsync(log3);

        // Act
        var results = await repository.FindByEntityAsync("Account", "1100");

        // Assert
        results.Should().HaveCount(3);
        results[0].Action.Should().Be(AuditAction.DELETE);  // 最新
        results[1].Action.Should().Be(AuditAction.UPDATE);
        results[2].Action.Should().Be(AuditAction.CREATE);  // 最古
    }
}

File: app/fsharp/AccountingSystem.Tests/Infrastructure/AuditLogRepositoryTest.fs
module AccountingSystem.Tests.Infrastructure.AuditLogRepositoryTest

open System
open AccountingSystem.Domain.Types
open AccountingSystem.Domain.Models
open AccountingSystem.Infrastructure.Persistence.Repositories
open AccountingSystem.Tests.DatabaseTestBase
open Npgsql
open Xunit
open FsUnit.Xunit

/// <summary>
/// 監査ログリポジトリ - Dapper 統合テスト
/// </summary>
type AuditLogRepositoryTest() =
    inherit DatabaseTestBase()

    /// テスト用の監査ログを作成
    member private _.CreateTestAuditLog
        (entityType: string)
        (entityId: string)
        (action: AuditAction)
        (userId: string)
        (userName: string)
        (timestamp: DateTime) : AuditLog =
        {
            Id = None
            EntityType = entityType
            EntityId = entityId
            Action = action
            UserId = userId
            UserName = userName
            Timestamp = timestamp
            OldValues = Some "{\"name\": \"old\"}"
            NewValues = Some "{\"name\": \"new\"}"
            Changes = Some "{\"name\": [\"old\", \"new\"]}"
            Reason = Some "テスト理由"
            IpAddress = Some "192.168.1.1"
            UserAgent = Some "TestAgent/1.0"
        }

    /// テストデータをクリーンアップ
    member private this.CleanupTestDataAsync() =
        task {
            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()

            use cmd = new NpgsqlCommand("""
                DELETE FROM audit_log WHERE entity_id LIKE 'TEST%';
            """, conn)
            let! _ = cmd.ExecuteNonQueryAsync()
            ()
        }

    [<Fact>]
    member this.``監査ログを登録できる``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let auditLog = this.CreateTestAuditLog
                            "Account" "TEST001" AuditAction.Create
                            "user1" "テストユーザー1" DateTime.UtcNow

            let! inserted = repository.InsertAsync(auditLog)

            inserted.Id.IsSome |> should equal true
            inserted.EntityType |> should equal "Account"
            inserted.EntityId |> should equal "TEST001"
            inserted.Action |> should equal AuditAction.Create

            do! this.CleanupTestDataAsync()
        }

    [<Fact>]
    member this.``エンティティ別の監査ログを取得できる``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            // 同じエンティティに対して複数のログを登録
            let log1 = this.CreateTestAuditLog
                        "Account" "TEST002" AuditAction.Create
                        "user1" "ユーザー1" (DateTime(2024, 1, 15, 10, 0, 0))
            let log2 = this.CreateTestAuditLog
                        "Account" "TEST002" AuditAction.Update
                        "user1" "ユーザー1" (DateTime(2024, 1, 16, 10, 0, 0))
            let log3 = this.CreateTestAuditLog
                        "Journal" "TEST003" AuditAction.Create
                        "user2" "ユーザー2" (DateTime(2024, 1, 17, 10, 0, 0))

            let! _ = repository.InsertAsync(log1)
            let! _ = repository.InsertAsync(log2)
            let! _ = repository.InsertAsync(log3)

            let! logs = repository.FindByEntityAsync("Account", "TEST002")

            logs.Length |> should equal 2
            logs |> List.forall (fun l -> l.EntityType = "Account" && l.EntityId = "TEST002")
                |> should equal true

            do! this.CleanupTestDataAsync()
        }

    [<Fact>]
    member this.``ユーザー別の監査ログを期間指定で取得できる``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let log1 = this.CreateTestAuditLog
                        "Account" "TEST004" AuditAction.Create
                        "testuser1" "テストユーザー1" (DateTime(2024, 1, 10, 10, 0, 0))
            let log2 = this.CreateTestAuditLog
                        "Account" "TEST005" AuditAction.Update
                        "testuser1" "テストユーザー1" (DateTime(2024, 1, 20, 10, 0, 0))
            let log3 = this.CreateTestAuditLog
                        "Journal" "TEST006" AuditAction.Create
                        "testuser2" "テストユーザー2" (DateTime(2024, 1, 15, 10, 0, 0))

            let! _ = repository.InsertAsync(log1)
            let! _ = repository.InsertAsync(log2)
            let! _ = repository.InsertAsync(log3)

            let! logs = repository.FindByUserAsync(
                            "testuser1",
                            DateTime(2024, 1, 1),
                            DateTime(2024, 1, 31))

            logs.Length |> should equal 2
            logs |> List.forall (fun l -> l.UserId = "testuser1")
                |> should equal true

            do! this.CleanupTestDataAsync()
        }

    [<Fact>]
    member this.``期間内のすべての監査ログを取得できる``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let log1 = this.CreateTestAuditLog
                        "Account" "TEST007" AuditAction.Create
                        "user1" "ユーザー1" (DateTime(2024, 1, 10, 10, 0, 0))
            let log2 = this.CreateTestAuditLog
                        "Account" "TEST008" AuditAction.Update
                        "user1" "ユーザー1" (DateTime(2024, 1, 15, 10, 0, 0))
            let log3 = this.CreateTestAuditLog
                        "Journal" "TEST009" AuditAction.Create
                        "user2" "ユーザー2" (DateTime(2024, 1, 25, 10, 0, 0))

            let! _ = repository.InsertAsync(log1)
            let! _ = repository.InsertAsync(log2)
            let! _ = repository.InsertAsync(log3)

            let! logs = repository.FindByDateRangeAsync(
                            DateTime(2024, 1, 12),
                            DateTime(2024, 1, 20))

            logs.Length |> should equal 1
            logs.Head.EntityId |> should equal "TEST008"

            do! this.CleanupTestDataAsync()
        }

    [<Fact>]
    member this.``アクション種別で監査ログを検索できる``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let log1 = this.CreateTestAuditLog
                        "Account" "TEST010" AuditAction.Create
                        "user1" "ユーザー1" (DateTime(2024, 1, 15, 10, 0, 0))
            let log2 = this.CreateTestAuditLog
                        "Account" "TEST011" AuditAction.Update
                        "user1" "ユーザー1" (DateTime(2024, 1, 16, 10, 0, 0))
            let log3 = this.CreateTestAuditLog
                        "Journal" "TEST012" AuditAction.Delete
                        "user2" "ユーザー2" (DateTime(2024, 1, 17, 10, 0, 0))

            let! _ = repository.InsertAsync(log1)
            let! _ = repository.InsertAsync(log2)
            let! _ = repository.InsertAsync(log3)

            let! createLogs = repository.FindByActionAsync(
                                "CREATE",
                                DateTime(2024, 1, 1),
                                DateTime(2024, 1, 31))

            createLogs.Length |> should equal 1
            createLogs.Head.Action |> should equal AuditAction.Create

            let! deleteLogs = repository.FindByActionAsync(
                                "DELETE",
                                DateTime(2024, 1, 1),
                                DateTime(2024, 1, 31))

            deleteLogs.Length |> should equal 1
            deleteLogs.Head.Action |> should equal AuditAction.Delete

            do! this.CleanupTestDataAsync()
        }

    [<Fact>]
    member this.``存在しないエンティティの監査ログは空リストを返す``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let! logs = repository.FindByEntityAsync("Account", "NONEXISTENT")

            logs.Length |> should equal 0
        }

    [<Fact>]
    member this.``JSONB形式のデータが正しく保存・取得できる``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let auditLog = {
                Id = None
                EntityType = "Account"
                EntityId = "TEST013"
                Action = AuditAction.Update
                UserId = "user1"
                UserName = "テストユーザー"
                Timestamp = DateTime.UtcNow
                OldValues = Some "{\"balance\": 10000, \"name\": \"旧名称\"}"
                NewValues = Some "{\"balance\": 20000, \"name\": \"新名称\"}"
                Changes = Some "{\"balance\": [10000, 20000], \"name\": [\"旧名称\", \"新名称\"]}"
                Reason = Some "残高修正"
                IpAddress = Some "10.0.0.1"
                UserAgent = Some "Mozilla/5.0"
            }

            let! inserted = repository.InsertAsync(auditLog)
            let! logs = repository.FindByEntityAsync("Account", "TEST013")

            logs.Length |> should equal 1
            let retrieved = logs.Head

            // JSONB は順序を保証しないため、値が存在することのみ確認
            retrieved.OldValues.IsSome |> should equal true
            retrieved.OldValues.Value.Contains("balance") |> should equal true
            retrieved.OldValues.Value.Contains("10000") |> should equal true
            retrieved.NewValues.IsSome |> should equal true
            retrieved.NewValues.Value.Contains("20000") |> should equal true
            retrieved.Changes.IsSome |> should equal true
            retrieved.Reason |> should equal (Some "残高修正")

            do! this.CleanupTestDataAsync()
        }

    [<Fact>]
    member this.``NULLのオプション値が正しく処理される``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let auditLog = {
                Id = None
                EntityType = "Account"
                EntityId = "TEST014"
                Action = AuditAction.Create
                UserId = "user1"
                UserName = "テストユーザー"
                Timestamp = DateTime.UtcNow
                OldValues = None
                NewValues = Some "{\"name\": \"新規作成\"}"
                Changes = None
                Reason = None
                IpAddress = None
                UserAgent = None
            }

            let! _ = repository.InsertAsync(auditLog)
            let! logs = repository.FindByEntityAsync("Account", "TEST014")

            logs.Length |> should equal 1
            let retrieved = logs.Head

            retrieved.OldValues |> should equal None
            retrieved.Changes |> should equal None
            retrieved.Reason |> should equal None
            retrieved.IpAddress |> should equal None
            retrieved.UserAgent |> should equal None

            do! this.CleanupTestDataAsync()
        }

    [<Fact>]
    member this.``監査ログは時系列順（降順）でソートされる``() =
        task {
            do! this.CleanupTestDataAsync()

            use conn = new NpgsqlConnection(this.ConnectionString)
            do! conn.OpenAsync()
            let repository = AuditLogRepository(conn)

            let log1 = this.CreateTestAuditLog
                        "Account" "TEST015" AuditAction.Create
                        "user1" "ユーザー1" (DateTime(2024, 1, 10, 10, 0, 0))
            let log2 = this.CreateTestAuditLog
                        "Account" "TEST015" AuditAction.Update
                        "user1" "ユーザー1" (DateTime(2024, 1, 20, 10, 0, 0))
            let log3 = this.CreateTestAuditLog
                        "Account" "TEST015" AuditAction.Update
                        "user2" "ユーザー2" (DateTime(2024, 1, 15, 10, 0, 0))

            let! _ = repository.InsertAsync(log1)
            let! _ = repository.InsertAsync(log2)
            let! _ = repository.InsertAsync(log3)

            let! logs = repository.FindByEntityAsync("Account", "TEST015")

            logs.Length |> should equal 3
            // 降順でソートされていることを確認
            logs.[0].Timestamp |> should be (greaterThan logs.[1].Timestamp)
            logs.[1].Timestamp |> should be (greaterThan logs.[2].Timestamp)

            do! this.CleanupTestDataAsync()
        }
